homeassistant/components/fan/reproduce_state.py

```diff
@@ -73,18 +82,39 @@ async def _async_reproduce_state(
     if state.state == STATE_ON:
         # The fan should be on
         if cur_state.state != STATE_ON:
-            # Turn on the fan at first
-            service_calls[SERVICE_TURN_ON] = service_data
-
-        for attr, service in ATTRIBUTES.items():
-            # Call services to adjust the attributes
-            if attr in state.attributes and not check_attr_equal(
-                state.attributes, cur_state.attributes, attr
-            ):
-                data = service_data.copy()
-                data[attr] = state.attributes[attr]
-                service_calls[service] = data
-
+            # Turn on the fan with all the speed and modes settings.
+            # The `turn_on` method will figure out in which mode to
+            # turn the fan on.
+            data = service_data.copy()
+            for attr in SPEED_AND_MODES_ATTRIBUTES:
+                if attr in state.attributes and state.attributes[attr] is not None:
+                    data[attr] = state.attributes[attr]
+            service_calls[SERVICE_TURN_ON] = data
+        else:
+            # The simple attributes are copied directly.
```

Comment:
Why is it ok for these to have `None` values but not the others? Naively, i just read the method for `def set_direction(self, direction: str) -> None:` and it doesn't appear to accept a non-value.

Reply:
If the fan does not support the feature, can the service call be omitted? Or do we require calling the set call with a None value in some cases?

Reply:
Technically, yes we can omit the service call if the feature is not supported. The reason why calling the services even when not supported is okay is because in their schema definition, the values are marked as optional: https://github.com/home-assistant/core/blob/dev/homeassistant/components/fan/__init__.py#L126

Your question had me wonder if I can rely on just the fact that, when changing one, the other _should_ turn to None (unless the integration is broken), but then I realized that this does not help us here because the `cur_state` is a snapshot at the beginning of the restore, not updated live, so we can't rely on a sequence such as off → on(percentage) → `state.preset_mode == cur_state.preset_mode == None`.

If instead of batching the service calls we would request to act immediately, that would be possible… but I don't think that's desired.